Implement StartTLS support for LDAP auth modules

LordAro opened the pull request:
I think I've covered everything that wants to be updated, do let me know if I missed something

I have to say, it's not exactly great that the codebase contains 3 separate LDAP connection objects. I'd definitely recommend merging in some fashon...

LordAro commented:
Anything more I can do to this to help get it mreged?

NeverMin commented:
Hi @LordAro , nice PR, maybe you need to update bin/otrs.CheckModules.pl, do you try?

LordAro commented:
Net::LDAP only requires IO::Socket::SSL for start_tls, which is already in the dependency list

Jockel007 commented:
Is the PR ready to merge?

It would be nice to have this new feature.

LordAro commented:
...There's a 4th LDAP implementation? Oh dear. Not only that, but CustomerUser reuses the config from AuthModule. Because that makes sense...

Anyway, rebased and implemented

LordAro commented:
pretty sure the CI error isn't me...

rajsardhara commented:
Got Compilation failed

LordAro commented:
Line 83 isn't one I've touched...

rajsardhara commented:
That's correct and neither I've..

LordAro commented:
Found a thing, and fixed it

Jockel007 commented:
@LordAro Thanks!

rajsardhara commented:
@LordAro everything works fine, I'm able to find customers from LDAP but I'm still getting this error msg, Not sure what does it mean
`Jun 14 02:08:49 support OTRS-CGI-60[24591]: [Error][Kernel::System::CustomerUser::new][Line:83]: Attempt to reload Kernel/System/CustomerUser/LDAP.pm aborted.#012Compilation failed in require at /opt/otrs/Kernel/System/Main.pm line 81.
Jun 14 02:08:49 support OTRS-CGI-60[24591]: [Error][Kernel::System::CustomerUser::new][Line:83]: Can't load backend module Kernel::System::CustomerUser::LDAP!
Jun 14 02:08:49 support OTRS-CGI-60[24591]: [Error][Kernel::System::CustomerUser::new][Line:83]: Attempt to reload Kernel/System/CustomerUser/LDAP.pm aborted.#012Compilation failed in require at /opt/otrs/Kernel/System/Main.pm line 81.
Jun 14 02:08:49 support OTRS-CGI-60[24591]: [Error][Kernel::System::CustomerUser::new][Line:83]: Can't load backend module Kernel::System::CustomerUser::LDAP!
`

rajsardhara commented:
This is what I'm using in my Config.pm

`$Self->{CustomerUser} = {
    Name => 'LDAP Data Source',
    Module => 'Kernel::System::CustomerUser::LDAP',
    Params => {
        Host => 'la.xyz.net',
        BaseDN => 'dc=xyz,dc=net',
        SSCOPE => 'sub',
        UserDN => 'CN=Test,CN=Users,DC=xyz,DC=net',
        UserPw => '@12345',
        Params => {
                port => 389,
                timeout => 120,
                async => 0,
                version => 3,
            },
    },
    # customer unique id
    CustomerKey => 'sAMAccountName',
    CustomerID => 'mail',
    CustomerUserListFields => ['cn', 'mail'],
    CustomerUserSearchFields => ['sAMAccountName', 'cn', 'mail'],
    CustomerUserSearchPrefix => '',
    CustomerUserSearchSuffix => '*',
    CustomerUserSearchListLimit => 250,
    CustomerUserPostMasterSearchFields => ['mail'],
    CustomerUserNameFields => ['givenname', 'sn'],
    Map => [
        [ 'UserFirstname',  'Firstname',  'givenname',       1, 1, 'var', '', 0 ],
        [ 'UserLastname',   'Lastname',   'sn',              1, 1, 'var', '', 0 ],
        [ 'UserLogin',      'Username',   'sAMAccountName',             1, 1, 'var', '', 0 ],
        [ 'UserEmail',      'Email',      'mail',            1, 1, 'var', '', 0 ],
        [ 'UserCustomerID', 'CustomerID', 'mail',            0, 1, 'var', '', 0 ],
        [ 'UserPhone',      'Phone',      'telephonenumber', 1, 0, 'var', '', 0 ],
    ],
};`

Jockel007 commented:
Can this commit be pulled into OTRS:master?

It is a great feature to enable a secure communication through the standard ldap port 389. 
Currently only a non-standard port 636(ldaps) can be used für secure ldap communication.
